### ISSUE TITLE
[CHA-RL3c][RoomLifecycle] Redundant `_releaseInProgress` check

### ISSUE THREAD

**Issue by sacOO7** [ably/ably-chat-js#399]
- CHA-RL3c mentions -> If the room is in the RELEASING status, then the operation will return the result of the pending release operation, or throw any exception that it throws
- This has been implemented as a part of https://github.com/ably/ably-chat-js/blob/0e1b52336a9e15c711d62848a0c3ba841dc6ae77/src/core/room-lifecycle-manager.ts#L799-L818
- But this kinda conflicts with, CHA-RL7 which mentions each operation is mutually exclusive and next operation will only start when previous one finishes.
- That means, we don't need to worry about ongoing `release` op, since [new release operation will only start after the previous `release` operation has finished](https://github.com/ably/ably-chat-js/blob/0e1b52336a9e15c711d62848a0c3ba841dc6ae77/src/core/room-lifecycle-manager.ts#L786).
- This will eliminate the above check along with the `_releaseInProgress` flag, wdyt

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/CHA-725) by [Unito](https://www.unito.io)

**Comment by sacOO7** [comment]
Maybe `_releaseInProgress` was supposed to serve boolean check for previous `rooms.get` impl. New async `rooms.get` don't need it anymore wdyt

**Comment by AndyTWF** [comment]
Perhaps the spec could be a bit clearer - but I don't think they necessarily conflict here. The spec mentions that the operation will not "commence" until the previous has finished, however - it's quite possible that you could argue that checking whether an operation is already in progress and piggybacking off it could occur outside of this.

In fact for some future changes, we might want to make it so that simple checks on room status etc (i.e. the releasing check) could occur outside of the mutually exclusive zone. The only thing that really needs to be mutually exclusive is the part when we start doing channel ops.

`_releaseInProgress` - I'm not sure why we have it as we could just check for the room status being `releasing` tbh, it didn't come from anything related to async room getting, or before.

**Comment by sacOO7** [comment]
Also, it seems `this._lifecycle.status === RoomStatus.Releasing` will always return false inside `this._mtx.runExclusive` block

**Comment by AndyTWF** [comment]
> Also, it seems `this._lifecycle.status === RoomStatus.Releasing` will always return false inside `this._mtx.runExclusive` block, because previous releasing operation will always finish with `RoomStatus.Released` before next one can start

That is true, though I can imagine a world whereby depending on exact implementation, we might want to check the status before we request the mtx, and again once we have it (at least in JS where we have single thread execution).

**Comment by sacOO7** [comment]
I strongly feel that, exposing a new world without mtx, would compromise the integrity of the current roomLifeCycelManager where we proudly say each operation is `atomic`

Currently, every operation is executed under `mtx` ( as per CHA-RL7 ), which should remain the same in the future as well, so that there won't be race conditions in other SDKs. There will never be issues with ably-js since it's single-threaded anyway.

Also, having `mtx` simplifies implementation, like currently we can get rid of `_releaseInProgress` along with related impl. because `mtx` makes sure a new operation starts when the previous one finishes. This way implementation is easier to read and interpret.

Also, I don't feel there is any performance overhead as such, `mtx` will return a chained promise exposing the same result anyway 💁‍♂️

**Comment by sacOO7** [comment]
I feel we should never open that world, to ensure the atomicity of each operation and for the sake of other SDKs where we need to deal with concurrency issues due to multithreaded env 😬

**Comment by AndyTWF** [comment]
I agree that at the moment I don't think there's a strong reason to change what we have - I was just giving an example of how things could potentially change in the future if there was a use-case (is being able to check certain state values prematurely necessarily a bad thing, so long as you re-check them within the mutex, for example).

Re `_releaseInProgress` - we'll consider this for a future spec update to tidy this up a bit :)

**Comment by sacOO7** [comment]
Thanks, this is helpful 👍

**Comment by sacOO7** [comment]
As per https://github.com/ably/ably-chat-js/issues/397#issuecomment-2481060166, I believe we might like to remove the following checks as a part of `_doChannelWindDown`

1. `this._lifecycle.status === RoomStatus.Releasing` will always return `false`
https://github.com/ably/ably-chat-js/blob/533c4d0314bcce942a30b4d7d6a012c3b081420d/src/core/room-lifecycle-manager.ts#L642-L643
and 
2. `this._lifecycle.status !== RoomStatus.Releasing` will always return `true`
https://github.com/ably/ably-chat-js/blob/533c4d0314bcce942a30b4d7d6a012c3b081420d/src/core/room-lifecycle-manager.ts#L665-L666

Since every operation is atomic, it will reach a `deterministic` state. So, ideally, we should not have checks for `ongoing` operations.

**Comment by sacOO7** [comment]
We can get rid of a lot of redundant code since it is not needed.

**Comment by sacOO7** [comment]
Related spec https://github.com/ably/specification/pull/230 has been merged hence related code can be refactored/ completely removed : )
